lib/Runtime/Language/InterpreterStackFrame.cpp

```diff
@@ -8330,6 +8331,98 @@ const byte * InterpreterStackFrame::OP_ProfiledLoopBodyStart(const byte * ip)
         return args;
     }
 
+    HeapArgumentsObject * InterpreterStackFrame::CreateEmptyHeapArgumentsObject(ScriptContext* scriptContext)
+    {
+        HeapArgumentsObject * args = JavascriptOperators::CreateHeapArguments(this->function->GetRealFunctionObject(), this->m_inSlotsCount - 1, 0, nullptr, scriptContext);
+        this->m_arguments = args;
+        return args;
+    }
+
+    void InterpreterStackFrame::TrySetFrameObjectInHeapArgObj(ScriptContext * scriptContext, bool hasNonSimpleParams)
+    {
+        if (PHASE_OFF1(Js::StackArgFormalsOptPhase))
+        {
+            return;
+        }
+
+        ActivationObject * frameObject = (ActivationObject*)GetLocalClosure();
+        uint32 formalsCount = this->m_functionBody->GetInParamsCount() - 1;
+        Js::PropertyIdArray * propIds = nullptr;
+        Js::HeapArgumentsObject* heapArgObj = nullptr;
+
+        if (m_arguments != nullptr)
+        {
+            //We always set the Frame object to nullptr in BailOutRecord::EnsureArguments for stack args optimization.
+            if (((Js::HeapArgumentsObject*)(m_arguments))->GetFrameObject() == nullptr)
+            {
+                heapArgObj = (Js::HeapArgumentsObject*)m_arguments;
+            }
+            else
+            {
+                Assert(false);
+                return;
+            }
+        }
+
+        bool isCachedScope = false;
+
+        //For Non-simple params, we don't have a scope object created.
+        if (this->m_functionBody->NeedScopeObjectForArguments(hasNonSimpleParams))
+        {
+                isCachedScope = m_functionBody->HasCachedScopePropIds();
+                if (isCachedScope)
+                {
+                    Js::DynamicType *literalType = nullptr;
+
+                    propIds = this->m_functionBody->GetFormalsPropIdArray();
```

Comment:
Done.

---
In reply to: [67794959](https://github.com/Microsoft/ChakraCore/pull/1135#discussion_r67794959) [](ancestors = 67794959)